pkgs/build-support/setup-hooks/make-binary-wrapper.sh

```diff
@@ -0,0 +1,274 @@
+# Assert that FILE exists and is executable
+#
+# assertExecutable FILE
+assertExecutable() {
+    local file="$1"
+    [[ -f "$file" && -x "$file" ]] || \
+        die "Cannot wrap '$file' because it is not an executable file"
+}
+
+# Generate a binary executable wrapper for wrapping an executable.
+# The binary is compiled from generated C-code using gcc.
+# makeBinaryWrapper EXECUTABLE OUT_PATH ARGS
+
+# ARGS:
+# --argv0       NAME    : set name of executed process to NAME
+#                         (otherwise it’s called …-wrapped)
+# --set         VAR VAL : add VAR with value VAL to the executable’s
+#                         environment
+# --set-default VAR VAL : like --set, but only adds VAR if not already set in
+#                         the environment
+# --unset       VAR     : remove VAR from the environment
+# --add-flags   FLAGS   : add FLAGS to invocation of executable
+
+# --prefix          ENV SEP VAL   : suffix/prefix ENV with VAL, separated by SEP
+# --suffix
+
+# To troubleshoot a binary wrapper after you compiled it,
+# use the `strings` command or open the binary file in a text editor.
+makeBinaryWrapper() {
+    assertExecutable "$1"
+    makeDocumentedCWrapper "$1" "${@:3}" | gcc -Os -x c -o "$2" -
+}
+
+# Syntax: wrapProgramBinary <PROGRAM> <MAKE-WRAPPER FLAGS...>
+wrapProgramBinary() {
+    local prog="$1"
+    local hidden
+
+    assertExecutable "$prog"
+
+    hidden="$(dirname "$prog")/.$(basename "$prog")"-wrapped
+    while [ -e "$hidden" ]; do
+      hidden="${hidden}_"
+    done
+    mv "$prog" "$hidden"
+    # Silence warning about unexpanded $0:
+    # shellcheck disable=SC2016
+    makeBinaryWrapper "$hidden" "$prog" --argv0 '$0' "${@:2}"
+}
+
+# Generate source code for the wrapper in such a way that the wrapper source code
+# will still be readable even after compilation
+# makeDocumentedCWrapper EXECUTABLE ARGS
+# ARGS: same as makeBinaryWrapper
+makeDocumentedCWrapper() {
+    local src docs
+    src=$(makeCWrapper "$@")
+    docs=$(documentationString "$src")
```

Comment:
I've actually had this thought myself. Depends on whether people want to see what happens "under the hood" or if they prefer seeing the higher level description.

If someone is debugging the difference between makeWrapper and makeBinaryWrapper for their program - being able to see the C code would definitely be an advantage.

Switching to embedding inputs (arguments) instead of outputs (C-code) could certainly simplify the code generation quite a bit, by removing the need for uses_-flags. Making the code look tidy is less important if it is not embedded in the executable.

That said, if someone actually needs to debug makeCWrapper, it is an advantage if the code is as easy to read as possible. So in that sense, there might still be an advantage to keep the code minimal/readable, even if it doesn't the binary size.

Reply:
> That said, if someone actually needs to debug the output of makeCWrapper, it is an advantage if the code is as easy to read as possible.

as you've noted in your opening post, you can still get at the code by running `makeCWrapper`. perhaps the best way is in between, something like
```
# the C code for this wrapper binary can be recreated with
nix-shell -p makeBinaryWrapper --run '
  makeCWrapper <programName> \
    --argv0 <argv0> \
    ... etc ...
'
```
debugging the actual *wrapper* probably isn't a common operation, we *do* however inspect wrappers for what they do quite a lot.